week3_cardgame/cardgame/src/components/card/CardHandler.jsx

```diff
@@ -0,0 +1,94 @@
+import React, { useEffect, useState } from 'react';
+import Card from './Card';
+import CARDLIST from './CardData';
+import styled from 'styled-components';
+
+export default function CardHandler({ level, updateScore, shuffle }) {
+    const [cards, setCards] = useState([]);
+    const [selectedCard, setSelectedCard] = useState(null); // 전에 선택한 카드
+    const [matchCards, setMatchCards] = useState([]); // 매칭된 카드들
+    const [openCards, setOpenCards] = useState([]); // 오픈된 카드들 (선택한 카드 + 매칭된 카드들)
+
+    // 랜덤 카드 뽑기
+    const getRandomElement = (arr) => {
+        console.log('getRandomElement');
+        let result = new Set();
+        while (result.size < level) {
+            const randomElement = arr[Math.floor(Math.random() * arr.length)];
+            result.add(randomElement);
+        }
+        return Array.from(result);
+    };
+
+    // 카드 쌍 만들기
+    const initialCards = () => {
+        const selectedCards = getRandomElement(CARDLIST);
+        const cardPairs = selectedCards.reduce(
+            (acc, card) => [
+                ...acc,
+                { ...card, id: card.id + 1000 }, // ID 충돌 방지
+                { ...card, id: card.id + 2000 }, // ID 충돌 방지
+            ],
+            [],
+        );
+        return cardPairs.sort(() => 0.5 - Math.random());
+    };
+
+    useEffect(() => {
+        setCards(initialCards());
+        setSelectedCard(null);
+        setMatchCards([]);
+        setOpenCards([]);
+    }, [level, shuffle]);
+
+    function chooseCard(e, id) {
+        e.stopPropagation();
+        // 매칭된 카드와 자기 자신은 선택 불가
+        if (!matchCards.includes(id) && id !== selectedCard) {
+            setOpenCards((prev) => [...prev, id]); // 카드 오픈
+            if (selectedCard == null) {
+                // 선택한 카드가 없을 때
+                setSelectedCard(id); // 선택한 카드에 추가
+            } else if (selectedCard % 1000 !== id % 1000) {
+                // 카드가 매칭되지 않을 때
+                setTimeout(() => {
+                    // 1초 지연
+                    setOpenCards(matchCards); // 선택 카드 다시 닫기
+                    setSelectedCard(null); // 선택한 카드 비우기
+                }, 500);
+            } else {
+                //카드가 매칭되었을 때
+                setMatchCards((prev) => {
+                    const newMatchCards = [...prev, id, selectedCard];
+                    setOpenCards(newMatchCards); // 매칭 반영 후 오픈 고정
+                    return newMatchCards;
+                });
+                setSelectedCard(null); //선택한 카드 비우기
+                updateScore((currentScore) => currentScore + 1);
+            }
+        }
+    }
+
+    return (
+        <CardContainer>
+            {cards.map((card, index) => (
+                <CardWrapper key={index} onClick={(e) => chooseCard(e, card.id)}>
```

Comment:
key 값으로 map의 순회 index를 주는 것은 지양해야 합니다!!
idx가 변경되는 이벤트가 발생하면 렌더되는 컴포넌트에서 사용하는 state가 의도하지 않은 방식으로 바뀔 수도 있기 때문이에요!!

고유하게 사용할 마땅한 값이 없으면 다음과 같이 해주는 방법도 많이 사용합니다!!
```suggestion
                <CardWrapper key={`card-${index}`} onClick={(e) => chooseCard(e, card.id)}>
```

Reply:
오 이런 방법이 있다니 신세계입니다! 적용 완료했습니다~~ㅎㅎ